sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -249,8 +255,26 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
       // By placing `fileFormatReaderGeneratedMetadataColumns` before `partitionColumns` and
       // `fileConstantMetadataColumns` in the `outputAttributes` we make these row operations
       // simpler and more efficient.
-      val outputAttributes = readDataColumns ++ fileFormatReaderGeneratedMetadataColumns ++
-        partitionColumns ++ fileConstantMetadataColumns
+      val outputAttributes = readDataColumns ++ generatedMetadataColumns ++
+        partitionColumns ++ constantMetadataColumns
+
+      // The attribute references in the filters also have to be categorized as either constant
+      // or generated metadata attributes. Only data filters can contain metadata filters.
+      def categorizeFileSourceMetadataAttributesInFilters(
```

Comment:
This was the least intrusive / most generic way I could come up with.
The problem is: When the filters come in, the attributes are already bound. Whatever changes we make that end up in outputAttributes, have to be applied to the filters as well, if we do not want all filter attributes to stay as FileSourceMetadataAttribute.

Reply:
This looks a bit complicated. Can we keep the previous behavior and don't support metadata col filter at all? https://github.com/apache/spark/pull/39314/files#diff-089285f1484c1598cb2839b86b6a9e65b98ab5b30462aedc210fe4bbf44cae78L386

We can support it later in an individual PR.

Reply:
The key problem is actually in PartitioningAwareFileIndex: There we build file level filters, and when row_index is not thrown out of the filter list, we end up with an empty file list because the row index filter filters out all files.
Let me check if that was broken before, but I think this is an important problem to fix